### DISCUSSION TITLE
Supporting batch processing

### DISCUSSION THREAD

**Discussion by janko** [jeremyevans/sequel#2080] [General]
In our Rails app I would like to sometimes use Sequel instead of Active Record for running data migrations. The one feature I'm missing in Sequel is Active Record's [`#in_batches`](https://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-in_batches), which returns datasets representing a batch of records, which makes it easier to perform bulk operations without actually loading records into memory.

One classic example is backfilling a newly added column:

```rb
Record.in_batches do |batch|
  batch.update_all(new_column: "some_value")
end
```

From what I could tell, the closest Sequel offers is `#paged_each`, which is the equivalent of [`#find_each`](https://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-find_each). It could technically be used to implement the above:

```rb
Record.order(:id).select(:id).naked.paged_each(row_per_fetch: BATCH_SIZE).each_slice(BATCH_SIZE) do |batch_data|
  batch_ids = batch_data.map { |data| data[:id] }
  batch = Record.where(id: batch_ids)
  # ...
end
```

Since this isn't that efficient, Active Record 7.1 will [optimize](https://blog.kiprosh.com/rails-7-1-optimizes-active-record-batching-for-whole-table-iterations/) this using an approach that seems similar to Sequel's `:filter` strategy for `#paged_each`. Would you be open to a PR adding support for batch processing like the one Active Record 7.1 will have?

**Comment by jeremyevans** [comment]
It looks like this could be implemented using an option to `paged_each`, to yield each dataset instead of yielding each row?  I'd be open to that approach.

**Reply by jeremyevans** [reply]
Giving this some more thought, maybe instead of a `paged_each` option, we split the method, add `paged_datasets`, and have `paged_each` call that?  Also, we could change the model version of `paged_each` to use a filter approach instead of an offset approach, at least for the common case of a single numeric primary key.

One thing I'm not sure how it would work is that `paged_each` currently forces a transaction. However, for what ActiveRecord wants (separate batch updates), you explicitly don't want a transaction.  I think it may be safe in the ActiveRecord case if you assuming an auto incrementing primary key. However, that may not be safe in the general case.  A filter approach certainly isn't going to work correctly if you are changing the filter column (primary key) during an update.

Obviously, we could punt here, make `paged_datasets` not use a transaction, and make `paged_each` use the transaction.  Seems like `paged_datasets` would have some sharp edges, and people who used it directly would have to know what they are doing, but that's not a deal breaker.

What are your thoughts?

**Reply by jeremyevans** [reply]
I took a brief look today and splitting `paged_datasets` and `paged_each` doesn't appear feasible.  In the :filter strategy, you need the results from the last row to set the filter for the next dataset.  In both the :filter and normal cases, you need to know when a dataset no longer yields rows so you can stop yielding.

In terms of the API:

```ruby
Record.in_batches do |batch|
  batch.update_all(new_column: "some_value")
end
```

`batch` cannot be a dataset, because it would not allow feeding the information back.  I think ActiveRecord yields a different type of object here, but I don't like that complexity, since you would expect that `batch` operates like a dataset.

My best idea at this point is an extension that adds `paged_update` and `paged_delete`, which have the same API as `update` and `delete`, but issue multiple queries internally to do the update/deletion.  Looking at the ActiveRecord pull request, it does not appear to be as optimized as it could be.   It looks like it loads all primary key values for every record it is deleting, just to throw most away.  A more optimized query approach for updates:

```ruby
ds = Record
offset_ds = ds.order(:id).offset(1001)
first = nil
while last = offset_ds.get(:id)
  ds.where(id: first...last).update(...)
  first = last
end
ds.where(id: first...last).update(...)
```

For deletes, it is even simpler:

```ruby
ds = Record
offset_ds = ds.order(:id).offset(1001)
while last = offset_ds.get(:id)
  ds.where{id < last}.delete
end
ds.delete
```

For selects, we could add a similar approach as described for updates to `paged_each`, and maybe support a `transaction: false` option to not use a transaction.  We already support paged inserts via `import`/`multi_insert`.

Note that the optimizations above work for scalar primary keys.  We could fallback to the ` WHERE ((pkc1, pkc2) IN (...))` approach for composite primary keys. It's possible to optimize the composite primary key case, but significantly more challenging and probably not worth it.

Thoughts on the above?

**Reply by jeremyevans** [reply]
I added a `paged_update_delete` plugin in b4ce555e4d2e23cc0e17a8e16da74fa8e15b3233.  I did not try to support the composite key case, as I determined you needed to support composite key ordering.  Such a thing is possible, but I'm not sure it's worth it due to the complexity.

For `paged_each`, `DB[:table].order(:id).paged_each(strategy: :filter)` already optimizes SELECT queries appropriately, so all that would need is support to disable transactions.  I wonder if a general option to `Database#transaction` that just synchronized without actually opening a transaction may make sense.  There are few places in Sequel that may or may not use a transaction based on provided options, and they would be simplified by supporting such a feature directly in `Dataset#transaction`.

**Reply by janko** [reply]
Sorry, I didn't find time to reply the last few days. My specific use case wasn't bulk updates/deletes, but bulk inserts into another table. I was splitting some columns into a new table, so I used `INSERT INTO new_table (<columns>) SELECT <values> FROM current_table ...`, where the `SELECT` part would be batched to avoid overloading the database. But I need to do bulk updates & deletes as well, so I would be interested in a generic batch API if that'll be possible.

Active Record's `in_batches` does yield actual relation objects, 7.0 does an `id IN (<list of IDs>)` filter, while 7.1 does ranged queries by default for full-table batching. Yeah, Active Record grabs the list of all IDs of the batch, but it appears to only use the last one, and also uses the count to detect whether it's last batch. It seems a bit wasteful.

I like the idea of a `#paged_datasets` method, which would yield datasets instead of records. I don't know if it'll be able to share any behavior with `#paged_each`. I'm not sure about the requirement of transactions, I was imagining this wouldn't use transactions. In my case we have scalar auto-incrementing primary key (no composite primary keys or UUIDs), so maybe `#paged_datasets` method could just require that.

I was thinking `#paged_datasets` would primarily support generic ranged batching, similar to your examples for bulk update & delete. Maybe it could also offer `IN` filtering for non-scalar primary keys, while composite primary keys wouldn't be supported, at least for now.

**Reply by jeremyevans** [reply]
Thanks for the feedback.  It turns out that it is easy to split a `paged_datasets` method from `paged_update`, and doing so significantly simplifies the `paged_update` code.  So I can add that to the `paged_update_delete` plugin.

**Reply by janko** [reply]
That's great to hear 🙂 Looking at the new plugin, will the update/delete work if the dataset has less total records than the batch size?

**Reply by jeremyevans** [reply]
Yep.  It's part of the specs.  In both cases, it just doesn't add an extra filter in that case.

**Reply by janko** [reply]
My mistake, I overlooked the part after the loop 👍

**Reply by jeremyevans** [reply]
OK, I pushed that change: 802b5de6521e32f7238a011d2a722f95bf353d6c

**Reply by jeremyevans** [reply]
I added support for a :skip_transaction option to `paged_each`: 89cdc6175398294c46c0ad867f619b00f12777d5

**Reply by janko** [reply]
Looks great! ❤️ I did some benchmarks, and in my dataset it was ~3x faster than Active Record 7.0 batching, and maybe ~20% faster than upcoming Active Record 7.1 implementation.

Given that it now provides a `#paged_datasets` method, I'm wondering if something more generic like `paged_operations` would be a better name of the plugin. In my particular case I would use it for paged inserts, which is neither an update nor delete.

Maybe it would also make sense to use the `batching` terminology, since it doesn't share common logic with `#paged_each`. So, we could have `#batch_update`, `#batch_delete`, `#batches` and `#batch_size`. Just an idea 🙂

**Reply by jeremyevans** [reply]
I'm fine renaming it to `paged_operations`.  Can you send a pull request for that?

I don't want to change from `page` to `batch`, as they describe the same type of operation, and `page` is Sequel's normal terminology for that.  While the plugin does not share an implementation with `paged_each`, the point is the same, to handle operations on a large dataset using multiple queries that each operate on part of the dataset.
